gateway/field_presence.go

```diff
@@ -0,0 +1,109 @@
+package gateway
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"reflect"
+	"strings"
+
+	"github.com/golang/protobuf/protoc-gen-go/generator"
+	"google.golang.org/genproto/protobuf/field_mask"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+const fieldPresenceMetaKey = "field-paths"
+
+// PresenceAnnotator will parse the JSON input and then add the paths to the
+// metadata to be pulled from context later
+func PresenceAnnotator(ctx context.Context, req *http.Request) metadata.MD {
+	if req == nil {
+		return nil
+	}
+	if req.Method != "POST" && req.Method != "PUT" { // && req.Method != "PATCH"
+		return nil
+	}
+
+	// Read body of request then reset it to be read in future
+	body, err := ioutil.ReadAll(req.Body)
+	req.Body = ioutil.NopCloser(bytes.NewReader(body))
+	if err != nil {
+		return nil
+	}
+
+	paths := []string{}
+	var root interface{}
+	if err := json.Unmarshal(body, &root); err != nil {
+		return nil
+	}
+
+	queue := []pathItem{{node: root}}
+	for len(queue) > 0 {
+		// dequeue an item
+		item := queue[0]
+		queue = queue[1:]
+		if m, ok := item.node.(map[string]interface{}); ok {
+			// if the item is an object, then enqueue all of its children
+			for k, v := range m {
+				queue = append(queue, pathItem{path: append(item.path, generator.CamelCase(k)), node: v})
+			}
+		} else if len(item.path) > 0 {
+			// otherwise, it's a leaf node so print its path
+			paths = append(paths, strings.Join(item.path, "."))
+		}
+	}
+
+	md := make(metadata.MD)
+	key := fieldPresenceMetaKey
+	for _, path := range paths {
+		md[key] = append(md[key], path)
```

Comment:
what's this doing? it looks like you've got a slice of strings and your iterating the slice to make a slice of strings.

Could you just set the slice without iterating?

Reply:
... (_sheepishly_) yes, that's exactly what it's doing, and yes, it should just be directly set.

Reply:
I'm glad I understood it, but I'm sad I didn't get to learn a new nifty Go trick :)

Reply:
It's a low level obfuscatory trick. You're free to copy it if you'd like, but I'd advise against it.